### ISSUE TITLE
Ktor controller routing

### ISSUE THREAD

**Issue by brezinajn** [Kodein-Framework/Kodein-DI#206]
It would be nice to be able to specify base routes for kodein controllers. For example if your API has private and public parts, where every controller in _/private_ is covered with auth, afaik there is no way to force controller to be registred on the private route. Great work on the ktor integration by the way 👍

**Comment by romainbsl** [comment]
Hi, glad you appreciate the integration with Ktor. 
I'm not sure to follow what you'd like. Have you done this with pure Ktor configuration ? 

Otherwise, let me rephrase. 

taking the gist https://gist.github.com/romainbsl/b21c4536a33548292edf033347c4e66c

Would you like that:
1. Controllers routes are automatically covered by auth with some configuration, like:

```kotlin
class LowerCaseController(application: Application) : SecuredKodeinController(auth = "baiscAuth") {
    override fun Routing.installRoutes() {
        route("/protected") {
            get("/user/lower") {
                val principal = call.principal<UserIdPrincipal>()!!
                call.respondText("Hello ${principal.name.toLowerCase()}!")
            }
        }
    }
}
```

Asking for authentication when accessing `/protected/user/lower`.

2. The controllers are automatically bound under a `/protected` route that is define elsewhere with athenticatiton on it, like the following snippet:

```kotlin
// APP configuration somewhere
fun Application.conf() {
    // ...
    routing {
        authenticate("basicAuth") {
            route("/protected") {
            }
        }
    }
    // ...
}
// Protected Controller
class LowerCaseController(application: Application) : EmbbededKodeinController(parentPath = "/protected") {
    override fun Routing.installRoutes() {
        get("/user/lower") {
            val principal = call.principal<UserIdPrincipal>()!!
            call.respondText("Hello ${principal.name.toLowerCase()}!")
        }
    }
}
```

Making this route accessible from "/protected/user/lower" with authentication.

3. I'm not following you, in that case, can you provide some details ?

**Comment by brezinajn** [comment]
Sorry, reading back I see my phrasing wasn't very clear. My use case is the 2nd one.

**Comment by SalomonBrys** [comment]
Well, having the base route in the constructor is the same thing as having it in `installRoutes`.

As I understand it, the controller itself should not know its own base route.

With the current API, this could be done so:

```kotlin
class LowerCaseController(application: Application, val basePath: String) : AbstractKodeinController(application) {
    override fun Routing.installRoutes() {
        route(basePath) {
            get("/user/lower") {
                val principal = call.principal<UserIdPrincipal>()!!
                call.respondText("Hello ${principal.name.toLowerCase()}!")
            }
        }
    }
}
```

```kotlin
const val PROTECTED_PATH = "/protected"

fun Application.conf() {
    routing {
        authenticate("basicAuth") {
            route("PROTECTED_PATH") {}
        }
    }
    kodein {
        bind() from singleton { LowerCaseController(instance(), PROTECTED_PATH) }
    }
    install(KodeinControllerFeature)
```

BUT

I have an idea.
Because controllers are always singletons, we could go berserker on the DSL path and create what's needed for this:

```kotlin
const val PROTECTED_PATH = "/protected"

fun Application.conf() {
    routing {
        authenticate("basicAuth") {
            route("PROTECTED_PATH") {}
        }
    }

    kodein {
        bind<MyService>() with singleton { MyImplementation() }

        controllers {
            controller { RootController(instance()) }
            route("/protected") {
                controller { ProtectedController(instance()) }
            }
        }
    }

    install(KodeinControllerFeature)
}
```

Ain't it fun?

**Comment by romainbsl** [comment]
Yes you're right, this can be done with the current API, but was not sure of the suggestion, so I tried to illustrate it.

This a nice idea, but that is why I asked if @brezinajn has already done that in pure Ktor. Because I don't think that Ktor allow to define the authentication on a path somewhere and apply it on the same path elsewhere. In fact, I'm sure that won't work because I just tested it.

Not sure how for the moment, but we could provide something like: 
```kotlin
const val PROTECTED_PATH = "/protected"

fun Application.conf() {
    kodein {
        bind<MyService>() with singleton { MyImplementation() }

        controllers {
            controller { RootController(instance()) }
            authenticate("basicAuth") {
                route(PROTECTED_PATH) {
                    controller { ProtectedController(instance()) }
                }
            }
        }
    }
    
    install(KodeinControllerFeature)
}
```

OR, as the controllers are significatively bound to `KodeinControllerFeature` we could move the logic into it:

```kotlin
fun Application.conf() {
    kodein {
        bind<MyService>() with singleton { MyImplementation() }
    }

    install(KodeinControllerFeature) {
        controller { RootController(instance()) }
        authenticate("basicAuth") {
            route(PROTECTED_PATH) {
                controller { ProtectedController(instance()) }
            }
        }
    }
}
```

Or, just drop the `install(KodeinControllerFeature)` as the logic could done in the `controllers { ... }`method.

```kotlin
const val PROTECTED_PATH = "/protected"

fun Application.conf() {
    kodein {
        bind<MyService>() with singleton { MyImplementation() }

        controllers {
            controller { RootController(instance()) }
            authenticate("basicAuth") {
                route(PROTECTED_PATH) {
                    controller { ProtectedController(instance()) }
                }
            }
        }
    }
}
```

**Comment by brezinajn** [comment]
> This a nice idea, but that is why I asked if @brezinajn has already done that in pure Ktor. Because I don't think that Ktor allow to define the authentication on a path somewhere and apply it on the same path elsewhere. In fact, I'm sure that won't work because I just tested it.

That is not possible.

The `controllers{}` approach seems great (it actually is quite similar to my solution pre ktor integration). The only thing I would suggest is option to specify endpoint in the controller block.
```Kotlin
const val PROTECTED_PATH = "protected"
const val USER_ENDPOINT = "user"
const val CAMPAIGN_ENDPOINT = "campaign"

fun Application.conf() {
    kodein {
        bind<MyService>() with singleton { MyImplementation() }

        controllers {
            controller { RootController(instance()) }
            authenticate("basicAuth") {
                route(PROTECTED_PATH) {
                    controller("user") { UserController(instance()) }
                    controller("campaign") { CampaignController(instance()) }
                }
            }
        }
    }
}
```
Another thing is this presumes definition of the controller instance in the route itself. It would be nice to be able to easily use controllers defined in another ktor module.

**Comment by romainbsl** [comment]
> The only thing I would suggest is option to specify endpoint in the controller block.

Why would you want that ? While you can do it directly in the `controller#installRoutes()` function, like @SalomonBrys said:

```kotlin
class LowerCaseController(application: Application, val basePath: String) : AbstractKodeinController(application) {
    override fun Routing.installRoutes() {
        route(basePath) {
            get("/lower") {
                val principal = call.principal<UserIdPrincipal>()!!
                call.respondText("Hello ${principal.name.toLowerCase()}!")
            }
        }
    }
}
```

It could be use like that:

```kotlin
controller { UserController(instance(), "user") }
```

But, I don't see the point if you don't want to reuse your controller for multiple uses.

> Another thing is this presumes definition of the controller instance in the route itself. It would be nice to be able to easily use controllers defined in another ktor module.

Would you like to define `controllers { }` ? In every module in your app, similarly to the `routing` feature ? Indeed, as it's in the `kodein { }` block, its constrained to it, but we could think of an approach to externalize it. Without going further this could be done by using Kodein module feature.  But if we find a relevant and elegant way to make it better it will be taken into account.

**Comment by brezinajn** [comment]
> Why would you want that ? While you can do it directly in the controller#installRoutes() function

This introduces coupling between the endpoint and the controller and adds IMO unnecessary clutter to constructor. I can see we're in opinionated waters now :). If it's not something you are sure about. Don't worry. It can be easily resolved by an extension function.
```Kotlin
fun Route.controller<T: KodeinAbstractController>(endpoint: String, factory: ()->T) = run{
    route(endpoint){
        controller{ factory() }
    }
}
```
Don't you love the language? :)

> Would you like to define controllers { } ? In every module in your app, similarly to the routing feature ? 

Yes.

**Comment by romainbsl** [comment]
@brezinajn hope this will help

**Comment by brezinajn** [comment]
Beautiful :) 👍
